sound/soc/soc-core.c

```diff
@@ -2839,12 +2839,14 @@ static void snd_soc_unbind_card(struct snd_soc_card *card, bool unregister)
 		snd_soc_dapm_shutdown(card);
 		snd_soc_flush_all_delayed_work(card);
 
+		mutex_lock(&client_mutex);
 		/* remove all components used by DAI links on this card */
 		for_each_comp_order(order) {
 			for_each_card_rtds(card, rtd) {
 				soc_remove_link_components(card, rtd, order);
 			}
 		}
+		mutex_unlock(&client_mutex);
```

Comment:
are you sure this needs to be here and not below

soc_cleanup_card_resources
  soc_remove_dai_links
     soc_remove_links_components

Reply:
@plbossart in fact this lock needs to be held only for soc_remove_link_components() as this is when the topology removal happens. I've changed the code to acquire the lock only when the component is removed now. snd_soc_remove_dai_link() is what needs the client_mutex and this is called only when the topology is removed.

Reply:
humm, looks like I was commenting on stale code, not sure how this happened.
can you send this directly to alsa-devel to get feedback, and add a GitHub issue in the commit message if people want to see more details.